EVPN route type-5 gateway IP overlay Index

adharkar opened the pull request:
Draft https://tools.ietf.org/html/draft-ietf-bess-evpn-prefix-advertisement-11 explains the use of overlay indexes for recursive route resolution for EVPN RT-5.
This PR focuses on gateway IP overlay index. Resolution of gateway IP overlay index requires a type-2 route with its IP same as the gateway IP.

This PR has following commits:
1. CLI to advertise gateway IP with EVPN prefix route.
2. Route map set command to advertise gateway IP with EVPN prefix route.
3. Receive and parse evpn type-5 route having gateway IP overlay index.
4. Set the gateway IP overlay index as the nexthop of the route imported in the vrf.
5. Perform recursive resolution of the gateway IP. i.e. stitch the gateway IP with corresponding MAC/IP route.

mjstapp commented:
Are the Static Analysis warnings new to this PR? if so, please fix them

adharkar commented:
> We must have test changes and I see white space changes that need to be fixed up.

Ho Donald,
I have fixed whitespace issues in bgpd.h.
I will add basic topotest for this feature.

adharkar commented:
> Are the Static Analysis warnings new to this PR? if so, please fix them

Thanks for pointing out! I have fixed the warnings. Will fix if there are any further warnings.

adharkar commented:
Hi reviewers,
I have added basic topotests for this feature.
Please review the code and tests.
Thanks!

louberger commented:
valgrind info:

==16877== 24 bytes in 1 blocks are still reachable in loss record 1 of 7
==16877==    at 0x4C31B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==16877==    by 0x296D6F: qcalloc (memory.c:110)
==16877==    by 0x2928B2: listnode_new (linklist.c:44)
==16877==    by 0x2928B2: listnode_add (linklist.c:59)
==16877==    by 0x288033: hash_create_size (hash.c:62)
==16877==    by 0x185A48: bgp_evpn_init (bgp_evpn.c:6468)
==16877==    by 0x2007AA: bgp_create (bgpd.c:3056)
==16877==    by 0x2007AA: bgp_get (bgpd.c:3207)
==16877==    by 0x1E8760: bgp_get_vty (bgp_vty.c:419)
==16877==    by 0x1E8FCA: router_bgp (bgp_vty.c:1217)
==16877==    by 0x26E1BD: cmd_execute_command_real.isra.2 (command.c:909)
==16877==    by 0x2703A6: cmd_execute_command_strict (command.c:1015)
==16877==    by 0x2703A6: command_config_read_one_line (command.c:1164)
==16877==    by 0x2705EC: config_from_file (command.c:1222)
==16877==    by 0x2D1B8D: vty_read_file (vty.c:2302)
==16877==    by 0x2D1B8D: vty_read_config (vty.c:2524)
==16877== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:listnode_new
   fun:listnode_add
   fun:hash_create_size
   fun:bgp_evpn_init
   fun:bgp_create
   fun:bgp_get
   fun:bgp_get_vty
   fun:router_bgp
   fun:cmd_execute_command_real.isra.2
   fun:cmd_execute_command_strict
   fun:command_config_read_one_line
   fun:config_from_file
   fun:vty_read_file
   fun:vty_read_config
}
==16877== 34 bytes in 1 blocks are still reachable in loss record 3 of 7
==16877==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==16877==    by 0x61E1A29: strdup (strdup.c:42)
==16877==    by 0x296F9F: qstrdup (memory.c:122)
==16877==    by 0x288004: hash_create_size (hash.c:55)
==16877==    by 0x185A48: bgp_evpn_init (bgp_evpn.c:6468)
==16877==    by 0x2007AA: bgp_create (bgpd.c:3056)
==16877==    by 0x2007AA: bgp_get (bgpd.c:3207)
==16877==    by 0x1E8760: bgp_get_vty (bgp_vty.c:419)
==16877==    by 0x1E8FCA: router_bgp (bgp_vty.c:1217)
==16877==    by 0x26E1BD: cmd_execute_command_real.isra.2 (command.c:909)
==16877==    by 0x2703A6: cmd_execute_command_strict (command.c:1015)
==16877==    by 0x2703A6: command_config_read_one_line (command.c:1164)
==16877==    by 0x2705EC: config_from_file (command.c:1222)
==16877==    by 0x2D1B8D: vty_read_file (vty.c:2302)
==16877==    by 0x2D1B8D: vty_read_config (vty.c:2524)
==16877== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:malloc
   fun:strdup
   fun:qstrdup
   fun:hash_create_size
   fun:bgp_evpn_init
   fun:bgp_create
   fun:bgp_get
   fun:bgp_get_vty
   fun:router_bgp
   fun:cmd_execute_command_real.isra.2
   fun:cmd_execute_command_strict
   fun:command_config_read_one_line
   fun:config_from_file
   fun:vty_read_file
   fun:vty_read_config
}
==16877== 40 bytes in 1 blocks are still reachable in loss record 4 of 7
==16877==    at 0x4C31B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==16877==    by 0x296D6F: qcalloc (memory.c:110)
==16877==    by 0x28805C: hash_create_size (hash.c:60)
==16877==    by 0x2C7481: thread_master_create (thread.c:452)
==16877==    by 0x291F27: frr_init (libfrr.c:697)
==16877==    by 0x16229F: main (bgp_main.c:486)
==16877== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:hash_create_size
   fun:thread_master_create
   fun:frr_init
   fun:main
}
==16877== 64 bytes in 1 blocks are still reachable in loss record 5 of 7
==16877==    at 0x4C31B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==16877==    by 0x296D6F: qcalloc (memory.c:110)
==16877==    by 0x287FBC: hash_create_size (hash.c:48)
==16877==    by 0x185A48: bgp_evpn_init (bgp_evpn.c:6468)
==16877==    by 0x2007AA: bgp_create (bgpd.c:3056)
==16877==    by 0x2007AA: bgp_get (bgpd.c:3207)
==16877==    by 0x1E8760: bgp_get_vty (bgp_vty.c:419)
==16877==    by 0x1E8FCA: router_bgp (bgp_vty.c:1217)
==16877==    by 0x26E1BD: cmd_execute_command_real.isra.2 (command.c:909)
==16877==    by 0x2703A6: cmd_execute_command_strict (command.c:1015)
==16877==    by 0x2703A6: command_config_read_one_line (command.c:1164)
==16877==    by 0x2705EC: config_from_file (command.c:1222)
==16877==    by 0x2D1B8D: vty_read_file (vty.c:2302)
==16877==    by 0x2D1B8D: vty_read_config (vty.c:2524)
==16877==    by 0x290A48: frr_config_read_in (libfrr.c:885)
==16877== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:hash_create_size
   fun:bgp_evpn_init
   fun:bgp_create
   fun:bgp_get
   fun:bgp_get_vty
   fun:router_bgp
   fun:cmd_execute_command_real.isra.2
   fun:cmd_execute_command_strict
   fun:command_config_read_one_line
   fun:config_from_file
   fun:vty_read_file
   fun:vty_read_config
   fun:frr_config_read_in
}
==16877== 2,048 bytes in 1 blocks are still reachable in loss record 7 of 7
==16877==    at 0x4C31B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==16877==    by 0x296D6F: qcalloc (memory.c:110)
==16877==    by 0x287FD3: hash_create_size (hash.c:50)
==16877==    by 0x185A48: bgp_evpn_init (bgp_evpn.c:6468)
==16877==    by 0x2007AA: bgp_create (bgpd.c:3056)
==16877==    by 0x2007AA: bgp_get (bgpd.c:3207)
==16877==    by 0x1E8760: bgp_get_vty (bgp_vty.c:419)
==16877==    by 0x1E8FCA: router_bgp (bgp_vty.c:1217)
==16877==    by 0x26E1BD: cmd_execute_command_real.isra.2 (command.c:909)
==16877==    by 0x2703A6: cmd_execute_command_strict (command.c:1015)
==16877==    by 0x2703A6: command_config_read_one_line (command.c:1164)
==16877==    by 0x2705EC: config_from_file (command.c:1222)
==16877==    by 0x2D1B8D: vty_read_file (vty.c:2302)
==16877==    by 0x2D1B8D: vty_read_config (vty.c:2524)
==16877==    by 0x290A48: frr_config_read_in (libfrr.c:885)
==16877== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:hash_create_size
   fun:bgp_evpn_init
   fun:bgp_create
   fun:bgp_get
   fun:bgp_get_vty
   fun:router_bgp
   fun:cmd_execute_command_real.isra.2
   fun:cmd_execute_command_strict
   fun:command_config_read_one_line
   fun:config_from_file
   fun:vty_read_file
   fun:vty_read_config
   fun:frr_config_read_in
}

adharkar commented:
Hi @donaldsharp @qlyoung @pjdruddy @vivek-cumulus , I have updated topotests for this feature and it passed CI. Can you review this PR?

pjdruddy commented:
@adharkar If you update to fix the conflicts I'll review this.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/2cc71b362f670931f1d5e399efaaa3c5/raw/6ba2b8ca75550eb7d142feb13c237fb53ca8d515/cr_6695_1610491003.diff | git apply
```

</p>

<p>

```diff
diff --git a/bgpd/bgp_debug.h b/bgpd/bgp_debug.h
index e0d17ab38..24677628b 100644
--- a/bgpd/bgp_debug.h
+++ b/bgpd/bgp_debug.h
@@ -37,8 +37,8 @@
 #define DUMP_DETAIL   32
 
 /* RD + Prefix + Path-Id */
-#define BGP_PRD_PATH_STRLEN (PREFIX_STRLEN + RD_ADDRSTRLEN \
-			     + INET6_ADDRSTRLEN + 34)
+#define BGP_PRD_PATH_STRLEN                                                    \
+	(PREFIX_STRLEN + RD_ADDRSTRLEN + INET6_ADDRSTRLEN + 34)
 
 extern int dump_open;
 extern int dump_update;
@@ -176,13 +176,11 @@ extern bool bgp_debug_update(struct peer *peer, const struct prefix *p,
 extern bool bgp_debug_bestpath(struct bgp_dest *dest);
 extern bool bgp_debug_zebra(const struct prefix *p);
 
-extern const char *
-bgp_debug_rdpfxpath2str(afi_t afi, safi_t safi, const struct prefix_rd *prd,
-			union prefixconstptr pu, mpls_label_t *label,
-			uint32_t num_labels, int addpath_valid,
-			uint32_t addpath_id,
-			struct bgp_route_evpn *overlay_index, char *str,
-			int size);
+extern const char *bgp_debug_rdpfxpath2str(
+	afi_t afi, safi_t safi, const struct prefix_rd *prd,
+	union prefixconstptr pu, mpls_label_t *label, uint32_t num_labels,
+	int addpath_valid, uint32_t addpath_id,
+	struct bgp_route_evpn *overlay_index, char *str, int size);
 const char *bgp_notify_admin_message(char *buf, size_t bufsz, uint8_t *data,
 				     size_t datalen);
 
diff --git a/bgpd/bgp_evpn.c b/bgpd/bgp_evpn.c
index 24a26026a..24212c760 100644
--- a/bgpd/bgp_evpn.c
+++ b/bgpd/bgp_evpn.c
@@ -1352,19 +1352,19 @@ static int update_evpn_type5_route(struct bgp *bgp_vrf, struct prefix_evpn *evp,
 
 	attr.mp_nexthop_len = BGP_ATTR_NHLEN_IPV4;
 
-	if (src_afi == AFI_IP6 &&
-	    CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-		       BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP)) {
-		if (src_attr &&
-		    !IN6_IS_ADDR_UNSPECIFIED(&src_attr->mp_nexthop_global)) {
+	if (src_afi == AFI_IP6
+	    && CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+			  BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP)) {
+		if (src_attr
+		    && !IN6_IS_ADDR_UNSPECIFIED(&src_attr->mp_nexthop_global)) {
 			attr.evpn_overlay.type = OVERLAY_INDEX_GATEWAY_IP;
 			memcpy(&attr.evpn_overlay.gw_ip.ipv6,
 			       &src_attr->mp_nexthop_global,
 			       sizeof(struct in6_addr));
 		}
-	} else if (src_afi == AFI_IP &&
-		   CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-			      BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP)) {
+	} else if (src_afi == AFI_IP
+		   && CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+				 BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP)) {
 		if (src_attr && src_attr->nexthop.s_addr != 0) {
 			attr.evpn_overlay.type = OVERLAY_INDEX_GATEWAY_IP;
 			memcpy(&attr.evpn_overlay.gw_ip.ipv4,
@@ -2483,8 +2483,8 @@ static int install_evpn_route_entry_in_vrf(struct bgp *bgp_vrf,
 			zlog_debug(
 				"Install gateway IP %s as nexthop for prefix %pFX in vrf %s",
 				inet_ntop(pp->family, &attr.evpn_overlay.gw_ip,
-					  buf1, sizeof(buf1)), pp,
-					  vrf_id_to_name(bgp_vrf->vrf_id));
+					  buf1, sizeof(buf1)),
+				pp, vrf_id_to_name(bgp_vrf->vrf_id));
 		}
 
 		if (afi == AFI_IP6) {
@@ -2542,8 +2542,7 @@ static int install_evpn_route_entry_in_vrf(struct bgp *bgp_vrf,
 			bgp_path_info_set_flag(dest, pi, BGP_PATH_VALID);
 		else {
 			if (BGP_DEBUG(nht, NHT)) {
-				inet_ntop(pp->family,
-					  &attr.evpn_overlay.gw_ip,
+				inet_ntop(pp->family, &attr.evpn_overlay.gw_ip,
 					  buf1, sizeof(buf1));
 				zlog_debug("%s: gateway IP NH unresolved",
 					   buf1);
@@ -4090,22 +4089,24 @@ static int process_type5_route(struct peer *peer, afi_t afi, safi_t safi,
 	 * An update containing a non-zero gateway IP and a non-zero ESI
 	 * at the same time is should be treated as withdraw
 	 */
-	if (!is_zero_esi(&evpn.eth_s_id) &&
-	    !is_zero_gw_ip(&evpn.gw_ip, gw_afi)) {
-		flog_err(EC_BGP_EVPN_ROUTE_INVALID,
-			 "%s - Rx EVPN Type-5 ESI and gateway-IP both non-zero.",
-			 peer->host);
+	if (!is_zero_esi(&evpn.eth_s_id)
+	    && !is_zero_gw_ip(&evpn.gw_ip, gw_afi)) {
+		flog_err(
+			EC_BGP_EVPN_ROUTE_INVALID,
+			"%s - Rx EVPN Type-5 ESI and gateway-IP both non-zero.",
+			peer->host);
 		is_valid_update = false;
 	} else if (!is_zero_esi(&evpn.eth_s_id))
 		evpn.type = OVERLAY_INDEX_ESI;
 	else if (!is_zero_gw_ip(&evpn.gw_ip, gw_afi))
 		evpn.type = OVERLAY_INDEX_GATEWAY_IP;
 	if (attr) {
-		if (is_zero_mac(&attr->rmac) && is_zero_esi(&evpn.eth_s_id) &&
-		    is_zero_gw_ip(&evpn.gw_ip, gw_afi) && label == 0) {
-			flog_err(EC_BGP_EVPN_ROUTE_INVALID,
-				 "%s - Rx EVPN Type-5 ESI, gateway-IP, RMAC and label all zero",
-				 peer->host);
+		if (is_zero_mac(&attr->rmac) && is_zero_esi(&evpn.eth_s_id)
+		    && is_zero_gw_ip(&evpn.gw_ip, gw_afi) && label == 0) {
+			flog_err(
+				EC_BGP_EVPN_ROUTE_INVALID,
+				"%s - Rx EVPN Type-5 ESI, gateway-IP, RMAC and label all zero",
+				peer->host);
 			is_valid_update = false;
 		}
 
@@ -5215,10 +5216,9 @@ struct bgpevpn *bgp_evpn_lookup_vni(struct bgp *bgp, vni_t vni)
  * Create a new vpn - invoked upon configuration or zebra notification.
  */
 struct bgpevpn *bgp_evpn_new(struct bgp *bgp, vni_t vni,
-		struct in_addr originator_ip,
-		vrf_id_t tenant_vrf_id,
-		struct in_addr mcast_grp,
-		ifindex_t svi_ifindex)
+			     struct in_addr originator_ip,
+			     vrf_id_t tenant_vrf_id, struct in_addr mcast_grp,
+			     ifindex_t svi_ifindex)
 {
 	struct bgpevpn *vpn;
 
@@ -5724,10 +5724,8 @@ int bgp_evpn_local_vni_del(struct bgp *bgp, vni_t vni)
  * about are for the local-tunnel-ip and the (tenant) VRF.
  */
 int bgp_evpn_local_vni_add(struct bgp *bgp, vni_t vni,
-			   struct in_addr originator_ip,
-			   vrf_id_t tenant_vrf_id,
-			   struct in_addr mcast_grp,
-			   ifindex_t svi_ifindex)
+			   struct in_addr originator_ip, vrf_id_t tenant_vrf_id,
+			   struct in_addr mcast_grp, ifindex_t svi_ifindex)
 {
 	struct bgpevpn *vpn;
 	struct prefix_evpn p;
@@ -5753,9 +5751,10 @@ int bgp_evpn_local_vni_add(struct bgp *bgp, vni_t vni,
 			 * Unresolve all the gateway IP nexthops for this VNI
 			 * for old SVI
 			 */
-			hash_iterate(vpn->remote_ip_hash,
-				(void (*)(struct hash_bucket *,
-				void *))bgp_evpn_remote_ip_hash_unlink_nexthop,
+			hash_iterate(
+				vpn->remote_ip_hash,
+				(void (*)(struct hash_bucket *, void *))
+					bgp_evpn_remote_ip_hash_unlink_nexthop,
 				vpn);
 			bgp_evpn_unlink_from_vni_svi_hash(bgp, vpn);
 			vpn->svi_ifindex = svi_ifindex;
@@ -5765,9 +5764,10 @@ int bgp_evpn_local_vni_add(struct bgp *bgp, vni_t vni,
 			 * Resolve all the gateway IP nexthops for this VNI
 			 * for new SVI
 			 */
-			hash_iterate(vpn->remote_ip_hash,
-				(void (*)(struct hash_bucket *,
-				void *))bgp_evpn_remote_ip_hash_link_nexthop,
+			hash_iterate(
+				vpn->remote_ip_hash,
+				(void (*)(struct hash_bucket *, void *))
+					bgp_evpn_remote_ip_hash_link_nexthop,
 				vpn);
 		}
 
@@ -5778,9 +5778,10 @@ int bgp_evpn_local_vni_add(struct bgp *bgp, vni_t vni,
 			 * Unresolve all the gateway IP nexthops for this VNI
 			 * in old tenant vrf
 			 */
-			hash_iterate(vpn->remote_ip_hash,
-				(void (*)(struct hash_bucket *,
-				void *))bgp_evpn_remote_ip_hash_unlink_nexthop,
+			hash_iterate(
+				vpn->remote_ip_hash,
+				(void (*)(struct hash_bucket *, void *))
+					bgp_evpn_remote_ip_hash_unlink_nexthop,
 				vpn);
 			bgpevpn_unlink_from_l3vni(vpn);
 			vpn->tenant_vrf_id = tenant_vrf_id;
@@ -5790,9 +5791,10 @@ int bgp_evpn_local_vni_add(struct bgp *bgp, vni_t vni,
 			 * Resolve all the gateway IP nexthops for this VNI
 			 * in new tenant vrf
 			 */
-			hash_iterate(vpn->remote_ip_hash,
-				(void (*)(struct hash_bucket *,
-				void *))bgp_evpn_remote_ip_hash_link_nexthop,
+			hash_iterate(
+				vpn->remote_ip_hash,
+				(void (*)(struct hash_bucket *, void *))
+					bgp_evpn_remote_ip_hash_link_nexthop,
 				vpn);
 		}
 
@@ -5812,7 +5814,7 @@ int bgp_evpn_local_vni_add(struct bgp *bgp, vni_t vni,
 	/* Create or update as appropriate. */
 	if (!vpn) {
 		vpn = bgp_evpn_new(bgp, vni, originator_ip, tenant_vrf_id,
-				mcast_grp, svi_ifindex);
+				   mcast_grp, svi_ifindex);
 		if (!vpn) {
 			flog_err(
 				EC_BGP_VNI,
@@ -5934,9 +5936,8 @@ void bgp_evpn_init(struct bgp *bgp)
 {
 	bgp->vnihash =
 		hash_create(vni_hash_key_make, vni_hash_cmp, "BGP VNI Hash");
-	bgp->vni_svi_hash =
-		hash_create(vni_svi_hash_key_make, vni_svi_hash_cmp,
-			    "BGP VNI hash based on SVI ifindex");
+	bgp->vni_svi_hash = hash_create(vni_svi_hash_key_make, vni_svi_hash_cmp,
+					"BGP VNI hash based on SVI ifindex");
 	bgp->import_rt_hash =
 		hash_create(import_rt_hash_key_make, import_rt_hash_cmp,
 			    "BGP Import RT Hash");
@@ -6091,8 +6092,9 @@ static void bgp_evpn_remote_ip_hash_destroy(struct bgpevpn *vpn)
 		return;
 
 	hash_iterate(vpn->remote_ip_hash,
-	(void (*)(struct hash_bucket *, void *))bgp_evpn_remote_ip_hash_free,
-	vpn);
+		     (void (*)(struct hash_bucket *,
+			       void *))bgp_evpn_remote_ip_hash_free,
+		     vpn);
 
 	hash_free(vpn->remote_ip_hash);
 	vpn->remote_ip_hash = NULL;
@@ -6108,9 +6110,9 @@ static void bgp_evpn_remote_ip_hash_add(struct bgpevpn *vpn,
 
 	evp = (struct prefix_evpn *)&pi->net->p;
 
-	if (evp->family != AF_EVPN ||
-	    evp->prefix.route_type != BGP_EVPN_MAC_IP_ROUTE ||
-	    is_evpn_prefix_ipaddr_none(evp))
+	if (evp->family != AF_EVPN
+	    || evp->prefix.route_type != BGP_EVPN_MAC_IP_ROUTE
+	    || is_evpn_prefix_ipaddr_none(evp))
 		return;
 
 	tmp.addr = evp->prefix.macip_addr.ip;
@@ -6141,9 +6143,9 @@ static void bgp_evpn_remote_ip_hash_del(struct bgpevpn *vpn,
 
 	evp = (struct prefix_evpn *)&pi->net->p;
 
-	if (evp->family != AF_EVPN ||
-	    evp->prefix.route_type != BGP_EVPN_MAC_IP_ROUTE ||
-	    is_evpn_prefix_ipaddr_none(evp))
+	if (evp->family != AF_EVPN
+	    || evp->prefix.route_type != BGP_EVPN_MAC_IP_ROUTE
+	    || is_evpn_prefix_ipaddr_none(evp))
 		return;
 
 	tmp.addr = evp->prefix.macip_addr.ip;
@@ -6187,7 +6189,8 @@ void bgp_evpn_show_remote_ip_hash(struct hash_bucket *bucket, void *args)
 	struct vty *vty = (struct vty *)args;
 
 	vty_out(vty, "VNI: %u\n", vpn->vni);
-	hash_iterate(vpn->remote_ip_hash,
+	hash_iterate(
+		vpn->remote_ip_hash,
 		(void (*)(struct hash_bucket *, void *))show_remote_ip_entry,
 		vty);
 	vty_out(vty, "\n");
@@ -6342,13 +6345,13 @@ static void bgp_evpn_remote_ip_process_nexthops(struct bgpevpn *vpn,
 		afi = AFI_IP;
 		p.family = AF_INET;
 		memcpy(&(p.u.prefix4), &(addr->ipaddr_v4),
-			 sizeof(struct in_addr));
+		       sizeof(struct in_addr));
 		p.prefixlen = IPV4_MAX_BITLEN;
 	} else if (addr->ipa_type == IPADDR_V6) {
 		afi = AFI_IP6;
 		p.family = AF_INET6;
 		memcpy(&(p.u.prefix6), &(addr->ipaddr_v6),
-			 sizeof(struct in6_addr));
+		       sizeof(struct in6_addr));
 		p.prefixlen = IPV6_MAX_BITLEN;
 	} else
 		return;
@@ -6382,12 +6385,11 @@ static void bgp_evpn_remote_ip_process_nexthops(struct bgpevpn *vpn,
 		evaluate_paths(bnc);
 	}
 
-	 /* MAC/IP route or SVI or tenant vrf being deleted from EVI */
-	if (!resolve &&  bnc->flags & BGP_NEXTHOP_VALID) {
+	/* MAC/IP route or SVI or tenant vrf being deleted from EVI */
+	if (!resolve && bnc->flags & BGP_NEXTHOP_VALID) {
 		bnc->flags &= ~BGP_NEXTHOP_VALID;
 		bnc->flags |= BGP_NEXTHOP_EVPN_INCOMPLETE;
 		bnc->change_flags |= BGP_NEXTHOP_MACIP_CHANGED;
 		evaluate_paths(bnc);
 	}
 }
-
diff --git a/bgpd/bgp_evpn.h b/bgpd/bgp_evpn.h
index c4141665f..f91b81f10 100644
--- a/bgpd/bgp_evpn.h
+++ b/bgpd/bgp_evpn.h
@@ -63,18 +63,18 @@ static inline int advertise_type5_routes(struct bgp *bgp_vrf,
 	if (!bgp_vrf->l3vni)
 		return 0;
 
-	if ((afi == AFI_IP) &&
-	    ((CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-			 BGP_L2VPN_EVPN_ADV_IPV4_UNICAST)) ||
-	     (CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-			 BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP))))
+	if ((afi == AFI_IP)
+	    && ((CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+			    BGP_L2VPN_EVPN_ADV_IPV4_UNICAST))
+		|| (CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+			       BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP))))
 		return 1;
 
-	if ((afi == AFI_IP6) &&
-	    ((CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-			 BGP_L2VPN_EVPN_ADV_IPV6_UNICAST)) ||
-	     (CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-			 BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP))))
+	if ((afi == AFI_IP6)
+	    && ((CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+			    BGP_L2VPN_EVPN_ADV_IPV6_UNICAST))
+		|| (CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+			       BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP))))
 		return 1;
 
 	return 0;
diff --git a/bgpd/bgp_evpn_private.h b/bgpd/bgp_evpn_private.h
index b3ebe6748..07ce44643 100644
--- a/bgpd/bgp_evpn_private.h
+++ b/bgpd/bgp_evpn_private.h
@@ -626,10 +626,10 @@ extern void bgp_evpn_derive_auto_rd(struct bgp *bgp, struct bgpevpn *vpn);
 extern void bgp_evpn_derive_auto_rd_for_vrf(struct bgp *bgp);
 extern struct bgpevpn *bgp_evpn_lookup_vni(struct bgp *bgp, vni_t vni);
 extern struct bgpevpn *bgp_evpn_new(struct bgp *bgp, vni_t vni,
-		struct in_addr originator_ip,
-		vrf_id_t tenant_vrf_id,
-		struct in_addr mcast_grp,
-		ifindex_t svi_ifindex);
+				    struct in_addr originator_ip,
+				    vrf_id_t tenant_vrf_id,
+				    struct in_addr mcast_grp,
+				    ifindex_t svi_ifindex);
 extern void bgp_evpn_free(struct bgp *bgp, struct bgpevpn *vpn);
 extern bool bgp_evpn_lookup_l3vni_l2vni_table(vni_t vni);
 extern int update_routes_for_vni(struct bgp *bgp, struct bgpevpn *vpn);
diff --git a/bgpd/bgp_evpn_vty.c b/bgpd/bgp_evpn_vty.c
index 87172e71e..ea0a6a9cc 100644
--- a/bgpd/bgp_evpn_vty.c
+++ b/bgpd/bgp_evpn_vty.c
@@ -531,9 +531,9 @@ static void display_vni(struct vty *vty, struct bgpevpn *vpn, json_object *json)
 		else
 			json_object_string_add(json, "advertiseSviMacIp",
 					       "Disabled");
-		json_object_string_add(json, "sviInterface",
-				       ifindex2ifname(vpn->svi_ifindex,
-						      vpn->tenant_vrf_id));
+		json_object_string_add(
+			json, "sviInterface",
+			ifindex2ifname(vpn->svi_ifindex, vpn->tenant_vrf_id));
 	} else {
 		vty_out(vty, "VNI: %d", vpn->vni);
 		if (is_vni_live(vpn))
@@ -3607,15 +3607,13 @@ DEFUN_HIDDEN (no_bgp_evpn_advertise_vni_subnet,
 	return CMD_SUCCESS;
 }
 
-DEFUN (bgp_evpn_advertise_type5,
-       bgp_evpn_advertise_type5_cmd,
-       "advertise " BGP_AFI_CMD_STR "" BGP_SAFI_CMD_STR " [gateway-ip] [route-map WORD]",
-       "Advertise prefix routes\n"
-       BGP_AFI_HELP_STR
-       BGP_SAFI_HELP_STR
-       "advertise gateway IP overlay index\n"
-       "route-map for filtering specific routes\n"
-       "Name of the route map\n")
+DEFUN(bgp_evpn_advertise_type5, bgp_evpn_advertise_type5_cmd,
+      "advertise " BGP_AFI_CMD_STR "" BGP_SAFI_CMD_STR
+      " [gateway-ip] [route-map WORD]",
+      "Advertise prefix routes\n" BGP_AFI_HELP_STR BGP_SAFI_HELP_STR
+      "advertise gateway IP overlay index\n"
+      "route-map for filtering specific routes\n"
+      "Name of the route map\n")
 {
 	struct bgp *bgp_vrf = VTY_GET_CONTEXT(bgp); /* bgp vrf instance */
 	int idx_afi = 0;
@@ -3665,9 +3663,9 @@ DEFUN (bgp_evpn_advertise_type5,
 
 	if (afi == AFI_IP) {
 		if ((!CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-				 BGP_L2VPN_EVPN_ADV_IPV4_UNICAST)) &&
-		    (!CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-				 BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP))) {
+				 BGP_L2VPN_EVPN_ADV_IPV4_UNICAST))
+		    && (!CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+				    BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP))) {
 
 			/*
 			 * this is the case for first time ever configuration
@@ -3682,9 +3680,10 @@ DEFUN (bgp_evpn_advertise_type5,
 				SET_FLAG(
 					bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
 					BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP);
-		} else if ((oly == OVERLAY_INDEX_TYPE_NONE) &&
-			   (!CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-					BGP_L2VPN_EVPN_ADV_IPV4_UNICAST))) {
+		} else if ((oly == OVERLAY_INDEX_TYPE_NONE)
+			   && (!CHECK_FLAG(
+				   bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+				   BGP_L2VPN_EVPN_ADV_IPV4_UNICAST))) {
 
 			/*
 			 * This is modify case from gateway-ip
@@ -3695,10 +3694,10 @@ DEFUN (bgp_evpn_advertise_type5,
 				   BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP);
 			SET_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
 				 BGP_L2VPN_EVPN_ADV_IPV4_UNICAST);
-		} else if ((oly == OVERLAY_INDEX_GATEWAY_IP) &&
-			   (!CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-					BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP))
-			  ) {
+		} else if ((oly == OVERLAY_INDEX_GATEWAY_IP)
+			   && (!CHECK_FLAG(
+				   bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+				   BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP))) {
 
 			/*
 			 * This is modify case from no overlay index
@@ -3724,9 +3723,9 @@ DEFUN (bgp_evpn_advertise_type5,
 		}
 	} else {
 		if ((!CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-				 BGP_L2VPN_EVPN_ADV_IPV6_UNICAST)) &&
-		    (!CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-				BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP))) {
+				 BGP_L2VPN_EVPN_ADV_IPV6_UNICAST))
+		    && (!CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+				    BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP))) {
 
 			/*
 			 * this is the case for first time ever configuration
@@ -3741,9 +3740,10 @@ DEFUN (bgp_evpn_advertise_type5,
 				SET_FLAG(
 					bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
 					BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP);
-		} else if ((oly == OVERLAY_INDEX_TYPE_NONE) &&
-			   (!CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-					BGP_L2VPN_EVPN_ADV_IPV6_UNICAST))) {
+		} else if ((oly == OVERLAY_INDEX_TYPE_NONE)
+			   && (!CHECK_FLAG(
+				   bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+				   BGP_L2VPN_EVPN_ADV_IPV6_UNICAST))) {
 
 			/*
 			 * This is modify case from gateway-ip
@@ -3754,10 +3754,10 @@ DEFUN (bgp_evpn_advertise_type5,
 				   BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP);
 			SET_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
 				 BGP_L2VPN_EVPN_ADV_IPV6_UNICAST);
-		} else if ((oly == OVERLAY_INDEX_GATEWAY_IP) &&
-			   (!CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-					BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP))
-			  ) {
+		} else if ((oly == OVERLAY_INDEX_GATEWAY_IP)
+			   && (!CHECK_FLAG(
+				   bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+				   BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP))) {
 
 			/*
 			 * This is modify case from no overlay index
@@ -3767,7 +3767,7 @@ DEFUN (bgp_evpn_advertise_type5,
 			UNSET_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
 				   BGP_L2VPN_EVPN_ADV_IPV6_UNICAST);
 			SET_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-				   BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP);
+				 BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP);
 		} else {
 
 			/*
@@ -3852,9 +3852,9 @@ DEFUN (no_bgp_evpn_advertise_type5,
 		 * nothing to do
 		 */
 		if ((CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-				BGP_L2VPN_EVPN_ADV_IPV4_UNICAST)) ||
-		    (CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-				BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP))) {
+				BGP_L2VPN_EVPN_ADV_IPV4_UNICAST))
+		    || (CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+				   BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP))) {
 			bgp_evpn_withdraw_type5_routes(bgp_vrf, afi, safi);
 			UNSET_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
 				   BGP_L2VPN_EVPN_ADV_IPV4_UNICAST);
@@ -3867,9 +3867,9 @@ DEFUN (no_bgp_evpn_advertise_type5,
 		 * nothing to do
 		 */
 		if ((CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-				BGP_L2VPN_EVPN_ADV_IPV6_UNICAST)) ||
-		    (CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
-				BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP))){
+				BGP_L2VPN_EVPN_ADV_IPV6_UNICAST))
+		    || (CHECK_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+				   BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP))) {
 			bgp_evpn_withdraw_type5_routes(bgp_vrf, afi, safi);
 			UNSET_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
 				   BGP_L2VPN_EVPN_ADV_IPV6_UNICAST);
@@ -4145,10 +4145,7 @@ DEFUN(show_bgp_l2vpn_evpn_vni,
 DEFUN_HIDDEN(show_bgp_l2vpn_evpn_vni_remote_ip_hash,
 	     show_bgp_l2vpn_evpn_vni_remote_ip_hash_cmd,
 	     "show bgp l2vpn evpn vni remote-ip-hash",
-	     SHOW_STR
-	     BGP_STR
-	     L2VPN_HELP_STR
-	     EVPN_HELP_STR
+	     SHOW_STR BGP_STR L2VPN_HELP_STR EVPN_HELP_STR
 	     "Show VNI\n"
 	     "Remote IP hash\n")
 {
@@ -4163,8 +4160,9 @@ DEFUN_HIDDEN(show_bgp_l2vpn_evpn_vni_remote_ip_hash,
 		return CMD_WARNING;
 
 	hash_iterate(bgp_evpn->vnihash,
-	(void (*)(struct hash_bucket *, void *))bgp_evpn_show_remote_ip_hash,
-	vty);
+		     (void (*)(struct hash_bucket *,
+			       void *))bgp_evpn_show_remote_ip_hash,
+		     vty);
 
 	return CMD_SUCCESS;
 }
@@ -4172,10 +4170,7 @@ DEFUN_HIDDEN(show_bgp_l2vpn_evpn_vni_remote_ip_hash,
 DEFUN_HIDDEN(show_bgp_l2vpn_evpn_vni_svi_hash,
 	     show_bgp_l2vpn_evpn_vni_svi_hash_cmd,
 	     "show bgp l2vpn evpn vni-svi-hash",
-	     SHOW_STR
-	     BGP_STR
-	     L2VPN_HELP_STR
-	     EVPN_HELP_STR
+	     SHOW_STR BGP_STR L2VPN_HELP_STR EVPN_HELP_STR
 	     "Show vni-svi-hash\n")
 {
 	struct bgp *bgp_evpn;
@@ -4189,8 +4184,9 @@ DEFUN_HIDDEN(show_bgp_l2vpn_evpn_vni_svi_hash,
 		return CMD_WARNING;
 
 	hash_iterate(bgp_evpn->vni_svi_hash,
-	(void (*)(struct hash_bucket *, void *))bgp_evpn_show_vni_svi_hash,
-	vty);
+		     (void (*)(struct hash_bucket *,
+			       void *))bgp_evpn_show_vni_svi_hash,
+		     vty);
 
 	return CMD_SUCCESS;
 }
@@ -5975,10 +5971,9 @@ void bgp_config_write_evpn_info(struct vty *vty, struct bgp *bgp, afi_t afi,
 			vty_out(vty, "  advertise ipv4 unicast route-map %s\n",
 				bgp->adv_cmd_rmap[AFI_IP][SAFI_UNICAST].name);
 		else
-			vty_out(vty,
-				"  advertise ipv4 unicast\n");
+			vty_out(vty, "  advertise ipv4 unicast\n");
 	} else if (CHECK_FLAG(bgp->af_flags[AFI_L2VPN][SAFI_EVPN],
-		   BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP)) {
+			      BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP)) {
 		if (bgp->adv_cmd_rmap[AFI_IP][SAFI_UNICAST].name)
 			vty_out(vty,
 				"  advertise ipv4 unicast gateway-ip route-map %s\n",
@@ -5990,12 +5985,10 @@ void bgp_config_write_evpn_info(struct vty *vty, struct bgp *bgp, afi_t afi,
 	if (CHECK_FLAG(bgp->af_flags[AFI_L2VPN][SAFI_EVPN],
 		       BGP_L2VPN_EVPN_ADV_IPV6_UNICAST)) {
 		if (bgp->adv_cmd_rmap[AFI_IP6][SAFI_UNICAST].name)
-			vty_out(vty,
-				"  advertise ipv6 unicast route-map %s\n",
+			vty_out(vty, "  advertise ipv6 unicast route-map %s\n",
 				bgp->adv_cmd_rmap[AFI_IP6][SAFI_UNICAST].name);
 		else
-			vty_out(vty,
-				"  advertise ipv6 unicast\n");
+			vty_out(vty, "  advertise ipv6 unicast\n");
 	} else if (CHECK_FLAG(bgp->af_flags[AFI_L2VPN][SAFI_EVPN],
 			      BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP)) {
 		if (bgp->adv_cmd_rmap[AFI_IP6][SAFI_UNICAST].name)
diff --git a/bgpd/bgp_evpn_vty.h b/bgpd/bgp_evpn_vty.h
index 4a75c8932..afc111f39 100644
--- a/bgpd/bgp_evpn_vty.h
+++ b/bgpd/bgp_evpn_vty.h
@@ -26,7 +26,7 @@ extern void bgp_config_write_evpn_info(struct vty *vty, struct bgp *bgp,
 extern void bgp_ethernetvpn_init(void);
 
 #define L2VPN_HELP_STR        "Layer 2 Virtual Private Network\n"
-#define EVPN_HELP_STR         "Ethernet Virtual Private Network\n"
+#define EVPN_HELP_STR "Ethernet Virtual Private Network\n"
 
 int argv_find_and_parse_oly_idx(struct cmd_token **argv, int argc, int *oly_idx,
 				enum overlay_index_type *oly);
diff --git a/bgpd/bgp_mac.c b/bgpd/bgp_mac.c
index 7e4444d9a..f97a65a25 100644
--- a/bgpd/bgp_mac.c
+++ b/bgpd/bgp_mac.c
@@ -197,8 +197,8 @@ static void bgp_process_mac_rescan_table(struct bgp *bgp, struct peer *peer,
 					char pfx_buf[BGP_PRD_PATH_STRLEN];
 
 					bgp_debug_rdpfxpath2str(
-						AFI_L2VPN, SAFI_EVPN, &prd,
-						p, label_pnt, num_labels,
+						AFI_L2VPN, SAFI_EVPN, &prd, p,
+						label_pnt, num_labels,
 						pi->addpath_rx_id ? 1 : 0,
 						pi->addpath_rx_id, NULL,
 						pfx_buf, sizeof(pfx_buf));
diff --git a/bgpd/bgp_nexthop.c b/bgpd/bgp_nexthop.c
index 228f0254d..4ca8e4520 100644
--- a/bgpd/bgp_nexthop.c
+++ b/bgpd/bgp_nexthop.c
@@ -832,8 +832,7 @@ static void bgp_show_nexthop(struct vty *vty, struct bgp *bgp,
 			vty_out(vty, " EVPN Gateway IP");
 		vty_out(vty, "\n");
 		bgp_show_nexthops_detail(vty, bgp, bnc);
-	} else if (CHECK_FLAG(bnc->flags,
-			      BGP_NEXTHOP_EVPN_INCOMPLETE)) {
+	} else if (CHECK_FLAG(bnc->flags, BGP_NEXTHOP_EVPN_INCOMPLETE)) {
 		vty_out(vty,
 			" %s overlay index unresolved [IGP metric %d], #paths %d",
 			inet_ntop(bnc->prefix.family, &bnc->prefix.u.prefix,
diff --git a/bgpd/bgp_nexthop.h b/bgpd/bgp_nexthop.h
index 6bf032d82..17eea145d 100644
--- a/bgpd/bgp_nexthop.h
+++ b/bgpd/bgp_nexthop.h
@@ -80,14 +80,14 @@ struct bgp_nexthop_cache {
  * L3 unreachable | VALID      = 0 | VALID      = 0
  *                | INCOMPLETE = 0 | INCOMPLETE = 0
  */
-#define BGP_NEXTHOP_EVPN_INCOMPLETE   (1 << 7)
+#define BGP_NEXTHOP_EVPN_INCOMPLETE (1 << 7)
 
 	uint16_t change_flags;
 
 #define BGP_NEXTHOP_CHANGED           (1 << 0)
 #define BGP_NEXTHOP_METRIC_CHANGED    (1 << 1)
 #define BGP_NEXTHOP_CONNECTED_CHANGED (1 << 2)
-#define BGP_NEXTHOP_MACIP_CHANGED     (1 << 3)
+#define BGP_NEXTHOP_MACIP_CHANGED (1 << 3)
 
 	/* Back pointer to the cache tree this entry belongs to. */
 	struct bgp_nexthop_cache_head *tree;
diff --git a/bgpd/bgp_nht.c b/bgpd/bgp_nht.c
index 7d7c256d1..9b6795c7a 100644
--- a/bgpd/bgp_nht.c
+++ b/bgpd/bgp_nht.c
@@ -427,9 +427,11 @@ static void bgp_process_nexthop_update(struct bgp_nexthop_cache *bnc,
 				char buf2[PREFIX2STR_BUFFER];
 
 				prefix2str(&bnc->prefix, buf2, sizeof(buf2));
-				zlog_debug("EVPN gateway IP %s recursive MAC/IP lookup %s",
-					   buf2, (evpn_resolved ?
-						  "successful" : "failed"));
+				zlog_debug(
+					"EVPN gateway IP %s recursive MAC/IP lookup %s",
+					buf2,
+					(evpn_resolved ? "successful"
+						       : "failed"));
 			}
 
 			if (evpn_resolved) {
@@ -777,12 +779,10 @@ void evaluate_paths(struct bgp_nexthop_cache *bnc)
 		bool bnc_is_valid_nexthop = false;
 		bool path_valid = false;
 
-		if (safi == SAFI_UNICAST &&
-			path->sub_type == BGP_ROUTE_IMPORTED &&
-			path->extra &&
-			path->extra->num_labels &&
-			path->attr->evpn_overlay.type !=
-				OVERLAY_INDEX_GATEWAY_IP) {
+		if (safi == SAFI_UNICAST && path->sub_type == BGP_ROUTE_IMPORTED
+		    && path->extra && path->extra->num_labels
+		    && path->attr->evpn_overlay.type
+			       != OVERLAY_INDEX_GATEWAY_IP) {
 			bnc_is_valid_nexthop =
 				bgp_isvalid_labeled_nexthop(bnc) ? true : false;
 		} else {
diff --git a/bgpd/bgp_route.c b/bgpd/bgp_route.c
index 548b7a31b..fc4042f2c 100644
--- a/bgpd/bgp_route.c
+++ b/bgpd/bgp_route.c
@@ -9542,8 +9542,8 @@ void route_vty_out_detail(struct vty *vty, struct bgp *bgp,
 		}
 	}
 
-	if (safi == SAFI_EVPN &&
-	    attr->evpn_overlay.type == OVERLAY_INDEX_GATEWAY_IP) {
+	if (safi == SAFI_EVPN
+	    && attr->evpn_overlay.type == OVERLAY_INDEX_GATEWAY_IP) {
 		char gwip_buf[INET6_ADDRSTRLEN];
 
 		if (is_evpn_prefix_ipaddr_v4((struct prefix_evpn *)&bn->p))
diff --git a/bgpd/bgp_routemap.c b/bgpd/bgp_routemap.c
index d3d4360ea..3188f8b2b 100644
--- a/bgpd/bgp_routemap.c
+++ b/bgpd/bgp_routemap.c
@@ -1000,8 +1000,8 @@ route_set_evpn_gateway_ip(void *rule, const struct prefix *prefix, void *object)
 	if (evp->prefix.route_type != BGP_EVPN_IP_PREFIX_ROUTE)
 		return RMAP_OKAY;
 
-	if ((is_evpn_prefix_ipaddr_v4(evp) && IPADDRSZ(gw_ip) != 4) ||
-	    (is_evpn_prefix_ipaddr_v6(evp) && IPADDRSZ(gw_ip) != 16))
+	if ((is_evpn_prefix_ipaddr_v4(evp) && IPADDRSZ(gw_ip) != 4)
+	    || (is_evpn_prefix_ipaddr_v6(evp) && IPADDRSZ(gw_ip) != 16))
 		return RMAP_OKAY;
 
 	path = object;
@@ -4068,14 +4068,12 @@ DEFUN (no_match_evpn_default_route,
 				      RMAP_EVENT_MATCH_DELETED);
 }
 
-DEFUN (set_evpn_gw_ip_ipv4,
-       set_evpn_gw_ip_ipv4_cmd,
-       "set evpn gateway-ip ipv4 A.B.C.D",
-       SET_STR
-       EVPN_HELP_STR
-       "Set gateway IP for prefix advertisement route\n"
-       "IPv4 address\n"
-       "Gateway IP address in IPv4 format\n")
+DEFUN(set_evpn_gw_ip_ipv4, set_evpn_gw_ip_ipv4_cmd,
+      "set evpn gateway-ip ipv4 A.B.C.D",
+      SET_STR EVPN_HELP_STR
+      "Set gateway IP for prefix advertisement route\n"
+      "IPv4 address\n"
+      "Gateway IP address in IPv4 format\n")
 {
 	int ret;
 	union sockunion su;
@@ -4097,14 +4095,12 @@ DEFUN (set_evpn_gw_ip_ipv4,
 			       "evpn gateway-ip ipv4", argv[4]->arg);
 }
 
-DEFUN (set_evpn_gw_ip_ipv6,
-       set_evpn_gw_ip_ipv6_cmd,
-       "set evpn gateway-ip ipv6 X:X::X:X",
-       SET_STR
-       EVPN_HELP_STR
-       "Set gateway IP for prefix advertisement route\n"
-       "IPv6 address\n"
-       "Gateway IP address in IPv6 format\n")
+DEFUN(set_evpn_gw_ip_ipv6, set_evpn_gw_ip_ipv6_cmd,
+      "set evpn gateway-ip ipv6 X:X::X:X",
+      SET_STR EVPN_HELP_STR
+      "Set gateway IP for prefix advertisement route\n"
+      "IPv6 address\n"
+      "Gateway IP address in IPv6 format\n")
 {
 	int ret;
 	union sockunion su;
@@ -4126,15 +4122,12 @@ DEFUN (set_evpn_gw_ip_ipv6,
 			       "evpn gateway-ip ipv6", argv[4]->arg);
 }
 
-DEFUN (no_set_evpn_gw_ip_ipv4,
-       no_set_evpn_gw_ip_ipv4_cmd,
-       "no set evpn gateway-ip ipv4 A.B.C.D",
-       NO_STR
-       SET_STR
-       EVPN_HELP_STR
-       "Set gateway IP for prefix advertisement route\n"
-       "IPv4 address\n"
-       "Gateway IP address in IPv4 format\n")
+DEFUN(no_set_evpn_gw_ip_ipv4, no_set_evpn_gw_ip_ipv4_cmd,
+      "no set evpn gateway-ip ipv4 A.B.C.D",
+      NO_STR SET_STR EVPN_HELP_STR
+      "Set gateway IP for prefix advertisement route\n"
+      "IPv4 address\n"
+      "Gateway IP address in IPv4 format\n")
 {
 	int ret;
 	union sockunion su;
@@ -4156,15 +4149,12 @@ DEFUN (no_set_evpn_gw_ip_ipv4,
 				  "evpn gateway-ip ipv4", argv[5]->arg);
 }
 
-DEFUN (no_set_evpn_gw_ip_ipv6,
-       no_set_evpn_gw_ip_ipv6_cmd,
-       "no set evpn gateway-ip ipv6 X:X::X:X",
-       NO_STR
-       SET_STR
-       EVPN_HELP_STR
-       "Set gateway IP for prefix advertisement route\n"
-       "IPv6 address\n"
-       "Gateway IP address in IPv6 format\n")
+DEFUN(no_set_evpn_gw_ip_ipv6, no_set_evpn_gw_ip_ipv6_cmd,
+      "no set evpn gateway-ip ipv6 X:X::X:X",
+      NO_STR SET_STR EVPN_HELP_STR
+      "Set gateway IP for prefix advertisement route\n"
+      "IPv6 address\n"
+      "Gateway IP address in IPv6 format\n")
 {
 	int ret;
 	union sockunion su;
diff --git a/bgpd/bgp_zebra.c b/bgpd/bgp_zebra.c
index d07b17912..a147e4833 100644
--- a/bgpd/bgp_zebra.c
+++ b/bgpd/bgp_zebra.c
@@ -1400,8 +1400,9 @@ void bgp_zebra_announce(struct bgp_dest *dest, const struct prefix *p,
 			api_nh->labels[0] = label;
 		}
 
-		if (is_evpn &&
-		    mpinfo->attr->evpn_overlay.type != OVERLAY_INDEX_GATEWAY_IP)
+		if (is_evpn
+		    && mpinfo->attr->evpn_overlay.type
+			       != OVERLAY_INDEX_GATEWAY_IP)
 			memcpy(&api_nh->rmac, &(mpinfo->attr->rmac),
 			       sizeof(struct ethaddr));
 
@@ -2813,10 +2814,11 @@ static int bgp_zebra_process_local_vni(ZAPI_CALLBACK_ARGS)
 		return 0;
 
 	if (BGP_DEBUG(zebra, ZEBRA))
-		zlog_debug("Rx VNI %s VRF %s VNI %u tenant-vrf %s SVI ifindex %u",
-			   (cmd == ZEBRA_VNI_ADD) ? "add" : "del",
-			   vrf_id_to_name(vrf_id), vni,
-			   vrf_id_to_name(tenant_vrf_id), svi_ifindex);
+		zlog_debug(
+			"Rx VNI %s VRF %s VNI %u tenant-vrf %s SVI ifindex %u",
+			(cmd == ZEBRA_VNI_ADD) ? "add" : "del",
+			vrf_id_to_name(vrf_id), vni,
+			vrf_id_to_name(tenant_vrf_id), svi_ifindex);
 
 	if (cmd == ZEBRA_VNI_ADD)
 		return bgp_evpn_local_vni_add(
diff --git a/bgpd/bgpd.h b/bgpd/bgpd.h
index 9ad251aa2..6030092dc 100644
--- a/bgpd/bgpd.h
+++ b/bgpd/bgpd.h
@@ -478,20 +478,20 @@ struct bgp {
 
 	/* BGP Per AF flags */
 	uint16_t af_flags[AFI_MAX][SAFI_MAX];
-#define BGP_CONFIG_DAMPENING                    (1 << 0)
+#define BGP_CONFIG_DAMPENING (1 << 0)
 /* l2vpn evpn flags - 1 << 0 is used for DAMPENNG */
-#define BGP_L2VPN_EVPN_ADV_IPV4_UNICAST         (1 << 1)
-#define BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP   (1 << 2)
-#define BGP_L2VPN_EVPN_ADV_IPV6_UNICAST         (1 << 3)
-#define BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP   (1 << 4)
-#define BGP_L2VPN_EVPN_DEFAULT_ORIGINATE_IPV4   (1 << 5)
-#define BGP_L2VPN_EVPN_DEFAULT_ORIGINATE_IPV6   (1 << 6)
+#define BGP_L2VPN_EVPN_ADV_IPV4_UNICAST (1 << 1)
+#define BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP (1 << 2)
+#define BGP_L2VPN_EVPN_ADV_IPV6_UNICAST (1 << 3)
+#define BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP (1 << 4)
+#define BGP_L2VPN_EVPN_DEFAULT_ORIGINATE_IPV4 (1 << 5)
+#define BGP_L2VPN_EVPN_DEFAULT_ORIGINATE_IPV6 (1 << 6)
 /* import/export between address families */
-#define BGP_CONFIG_VRF_TO_MPLSVPN_EXPORT        (1 << 7)
-#define BGP_CONFIG_MPLSVPN_TO_VRF_IMPORT        (1 << 8)
+#define BGP_CONFIG_VRF_TO_MPLSVPN_EXPORT (1 << 7)
+#define BGP_CONFIG_MPLSVPN_TO_VRF_IMPORT (1 << 8)
 /* vrf-route leaking flags */
-#define BGP_CONFIG_VRF_TO_VRF_IMPORT            (1 << 9)
-#define BGP_CONFIG_VRF_TO_VRF_EXPORT            (1 << 10)
+#define BGP_CONFIG_VRF_TO_VRF_IMPORT (1 << 9)
+#define BGP_CONFIG_VRF_TO_VRF_EXPORT (1 << 10)
 
 	/* BGP per AF peer count */
 	uint32_t af_peer_count[AFI_MAX][SAFI_MAX];
diff --git a/zebra/zebra_evpn.c b/zebra/zebra_evpn.c
index ba06665e7..aa989ca90 100644
--- a/zebra/zebra_evpn.c
+++ b/zebra/zebra_evpn.c
@@ -1100,13 +1100,12 @@ int zebra_evpn_send_add_to_client(zebra_evpn_t *zevpn)
 	stream_putw_at(s, 0, stream_get_endp(s));
 
 	if (IS_ZEBRA_DEBUG_VXLAN)
-		zlog_debug("Send EVPN_ADD %u %pI4 tenant vrf %s(%u) SVI index %u to %s",
-			   zevpn->vni,
-			   &zevpn->local_vtep_ip,
-			   vrf_id_to_name(zevpn->vrf_id),
-			   zevpn->vrf_id,
-			   (zevpn->svi_if ? zevpn->svi_if->ifindex : 0),
-			   zebra_route_string(client->proto));
+		zlog_debug(
+			"Send EVPN_ADD %u %pI4 tenant vrf %s(%u) SVI index %u to %s",
+			zevpn->vni, &zevpn->local_vtep_ip,
+			vrf_id_to_name(zevpn->vrf_id), zevpn->vrf_id,
+			(zevpn->svi_if ? zevpn->svi_if->ifindex : 0),
+			zebra_route_string(client->proto));
 
 	client->vniadd_cnt++;
 	rc = zserv_send_message(client, s);
diff --git a/zebra/zebra_fpm_netlink.c b/zebra/zebra_fpm_netlink.c
index da5cae0b6..86c1d447a 100644
--- a/zebra/zebra_fpm_netlink.c
+++ b/zebra/zebra_fpm_netlink.c
@@ -213,8 +213,7 @@ static int netlink_route_info_add_nh(struct netlink_route_info *ri,
 					if (IS_ZEBRA_IF_BRIDGE(ifp))
 						link_if = ifp;
 					else if (IS_ZEBRA_IF_VLAN(ifp))
-						link_if =
-						if_lookup_by_index_per_ns(
+						link_if = if_lookup_by_index_per_ns(
 							zvrf->zns,
 							zif->link_ifindex);
 					if (link_if)
diff --git a/zebra/zebra_vxlan.c b/zebra/zebra_vxlan.c
index a93a8fb40..719c8c2e1 100644
--- a/zebra/zebra_vxlan.c
+++ b/zebra/zebra_vxlan.c
@@ -1857,8 +1857,7 @@ static zebra_l3vni_t *zl3vni_from_svi(struct interface *ifp,
 	return zl3vni;
 }
 
-vni_t vni_id_from_svi(struct interface *ifp,
-		      struct interface *br_if)
+vni_t vni_id_from_svi(struct interface *ifp, struct interface *br_if)
 {
 	vni_t vni = 0;
 	zebra_evpn_t *zevpn = NULL;

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

adharkar commented:
Hi @vivek-cumulus @qlyoung @pjdruddy @donaldsharp 
Can you review this PR?

I have addressed the comments and resolved the conflicts.

donaldsharp commented on doc/user/bgp.rst:
```diff
@@ -2717,6 +2717,92 @@ the same behavior of using same next-hop and RMAC values.
+1. CLI to add gateway IP while generating EVPN type-5 route from a BGP IPv4/IPv6
+prefix:
+
+.. index:: [no] advertise <ipv4|ipv6> unicast [gateway-ip]
```
remove the [no] from index lines

adharkar replied:
Resolved

donaldsharp commented on doc/user/bgp.rst:
```diff
@@ -2717,6 +2717,92 @@ the same behavior of using same next-hop and RMAC values.
+
+2. Add gateway IP to EVPN type-5 route using a route-map:
+
+.. index:: [no] set evpn gateway-ip <ipv4|ipv6> <addr>
```
again

adharkar replied:
Resolved

donaldsharp reviewed (changes requested):
I see *lots* of whitespace changes and bad formatting.  Please run *each* commit through something like clang-formatter.

Additionally there are still 3 SA issues being found.  Please address these like we have initiallly asked for

adharkar commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -3635,39 +3652,147 @@ DEFUN (bgp_evpn_advertise_type5,
 
+	if ((oly != OVERLAY_INDEX_TYPE_NONE)
+	    && (oly != OVERLAY_INDEX_GATEWAY_IP)) {
+		vty_out(vty, "%%Unknown overlay-index type speified");
```
Resolved

adharkar commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -3635,39 +3652,147 @@ DEFUN (bgp_evpn_advertise_type5,
+				SET_FLAG(
+					bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+					BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP);
+		} else if ((oly == OVERLAY_INDEX_TYPE_NONE) &&
```
Resolved

adharkar commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -3635,39 +3652,147 @@ DEFUN (bgp_evpn_advertise_type5,
+				   BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP);
+			SET_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+				 BGP_L2VPN_EVPN_ADV_IPV4_UNICAST);
+		} else if ((oly == OVERLAY_INDEX_GATEWAY_IP) &&
```
Resolved

adharkar commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -3635,39 +3652,147 @@ DEFUN (bgp_evpn_advertise_type5,
+				   BGP_L2VPN_EVPN_ADV_IPV6_UNICAST_GW_IP);
+			SET_FLAG(bgp_vrf->af_flags[AFI_L2VPN][SAFI_EVPN],
+				 BGP_L2VPN_EVPN_ADV_IPV6_UNICAST);
+		} else if ((oly == OVERLAY_INDEX_GATEWAY_IP) &&
```
Resolved

adharkar commented on bgpd/bgp_evpn_vty.h:
```diff
@@ -26,7 +26,10 @@ extern void bgp_config_write_evpn_info(struct vty *vty, struct bgp *bgp,
-#define EVPN_HELP_STR        "Ethernet Virtual Private Network\n"
+#define EVPN_HELP_STR         "Ethernet Virtual Private Network\n"
+
+int argv_find_and_parse_oly_idx(struct cmd_token **argv, int argc, int *oly_idx,
```
Resolved

adharkar commented on bgpd/bgp_evpn_vty.h:
```diff
@@ -26,7 +26,10 @@ extern void bgp_config_write_evpn_info(struct vty *vty, struct bgp *bgp,
 
 #define L2VPN_HELP_STR        "Layer 2 Virtual Private Network\n"
-#define EVPN_HELP_STR        "Ethernet Virtual Private Network\n"
+#define EVPN_HELP_STR         "Ethernet Virtual Private Network\n"
```
Resolved

adharkar commented on bgpd/bgpd.h:
```diff
@@ -478,18 +478,20 @@ struct bgp {
 
 	/* BGP Per AF flags */
 	uint16_t af_flags[AFI_MAX][SAFI_MAX];
-#define BGP_CONFIG_DAMPENING				(1 << 0)
```
Resolved

adharkar commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -4142,6 +4142,59 @@ DEFUN(show_bgp_l2vpn_evpn_vni,
 	return CMD_SUCCESS;
 }
 
+DEFUN_HIDDEN(show_bgp_l2vpn_evpn_vni_remote_ip_hash,
```
remote-ip-hash is an internal structure used for stitching type-2 routes to gateway IP. This show command will  mostly be used for debugging by developers and not by users. So I thought its more appropriate to make this hidden.

adharkar commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -4142,6 +4142,59 @@ DEFUN(show_bgp_l2vpn_evpn_vni,
+		return CMD_WARNING;
+
+	hash_iterate(bgp_evpn->vnihash,
+	(void (*)(struct hash_bucket *, void *))bgp_evpn_show_remote_ip_hash,
```
Resolved

adharkar commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -4142,6 +4142,59 @@ DEFUN(show_bgp_l2vpn_evpn_vni,
+
+DEFUN_HIDDEN(show_bgp_l2vpn_evpn_vni_svi_hash,
+	     show_bgp_l2vpn_evpn_vni_svi_hash_cmd,
+	     "show bgp l2vpn evpn vni-svi-hash",
```
vni_svi_hash is an internal structure used to find BGP EVPN given an SVI index. This show command will mostly be used for debugging by developers and not by users. So I thought its more appropriate to make this hidden.

adharkar commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -4142,6 +4142,59 @@ DEFUN(show_bgp_l2vpn_evpn_vni,
+		return CMD_WARNING;
+
+	hash_iterate(bgp_evpn->vni_svi_hash,
+	(void (*)(struct hash_bucket *, void *))bgp_evpn_show_vni_svi_hash,
```
Resolved

adharkar commented on bgpd/bgp_route.c:
```diff
@@ -9494,6 +9494,12 @@ void route_vty_out_detail(struct vty *vty, struct bgp *bgp,
+	if (safi == SAFI_EVPN) {
+		if (!json_paths)
+			vty_out(vty, "  Route %pFX",
+				(struct prefix_evpn *)bgp_dest_get_prefix(bn));
```
I quickly tested this and looks like it worked for EVPN type-3 route. But does it work for all EVPN prefixes? Code uses %pFX for EVPN prefixes at other places.

adharkar commented on bgpd/bgp_route.c:
```diff
@@ -9549,6 +9551,27 @@ void route_vty_out_detail(struct vty *vty, struct bgp *bgp,
 		}
 	}
 
+	if (safi == SAFI_EVPN &&
```
Resolved

adharkar replied:
Resolved

adharkar commented on bgpd/bgp_nht.c:
```diff
@@ -741,8 +744,9 @@ static void evaluate_paths(struct bgp_nexthop_cache *bnc)
 			path->extra &&
-			path->extra->num_labels) {
-
+			path->extra->num_labels &&
```
Resolved

adharkar commented on bgpd/bgp_evpn.c:
```diff
@@ -5716,13 +5747,53 @@ int bgp_evpn_local_vni_add(struct bgp *bgp, vni_t vni,
+			 */
+			hash_iterate(vpn->remote_ip_hash,
+				(void (*)(struct hash_bucket *,
+				void *))bgp_evpn_remote_ip_hash_link_nexthop,
```
What is the correct formatting for this block? bgp_evpn_remote_ip_hash_link_nexthop line goes beyond 80 characters if we try to align (void (*)(struct hash_bucket *, with (.

adharkar replied:
Resolved

adharkar commented on bgpd/bgp_evpn.c:
```diff
@@ -5950,3 +6026,368 @@ bool bgp_evpn_is_prefix_nht_supported(const struct prefix *pfx)
+	}
+
+	ip = hash_get(vpn->remote_ip_hash, &tmp, bgp_evpn_remote_ip_hash_alloc);
+	if (!ip)
```
Wouldn't this fail if somehow allocation failed?

adharkar commented on bgpd/bgp_evpn.c:
```diff
@@ -5950,3 +6026,368 @@ bool bgp_evpn_is_prefix_nht_supported(const struct prefix *pfx)
+	const struct bgpevpn *vpn1 = p1;
+	const struct bgpevpn *vpn2 = p2;
+
+	if (!vpn1 && !vpn2)
```
Resolved

adharkar commented on bgpd/bgp_evpn.c:
```diff
@@ -5950,3 +6026,368 @@ bool bgp_evpn_is_prefix_nht_supported(const struct prefix *pfx)
+	const struct evpn_remote_ip *ip1 = p1;
+	const struct evpn_remote_ip *ip2 = p2;
+
+	if (ip1 == NULL && ip2 == NULL)
```
Resolved

adharkar commented:
Hi @donaldsharp , I have addressed most of your comments. I will address remaining comments in next couple of days.
I ran checkpatch.sh for every commit. 

Regarding the static analysis errors, we see these errors because I have added following condition inside bgp_update function

if (attr && afi == AFI_L2VPN && evpn)

Looks like for this function some other paths assume that attr is always non-null. Is it safe to assume that?

adharkar commented on bgpd/bgp_attr_evpn.c:
```diff
@@ -315,3 +315,14 @@ extern bool is_zero_gw_ip(const union gw_addr *gw_ip, const afi_t afi)
+{
+	int i;
+
+	for (i = 0; i < 10; i++)
```
Resolved

adharkar commented:
Hi @donaldsharp I have addressed your comments.

Regarding the static analysis errors, we see these errors because I have added following condition inside bgp_update function

if (attr && afi == AFI_L2VPN && evpn)

Looks like for this function some other paths assume that attr is always non-null. Is it safe to assume that?

adharkar commented:
CI:rerun

adharkar commented:
CI:rerun

adharkar commented:
Hi @donaldsharp I have addressed your comments.

Regarding the static analysis errors, we see these errors because I have added following condition inside bgp_update function

if (attr && afi == AFI_L2VPN && evpn)

Looks like for this function some other paths assume that attr is always non-null. Is it safe to assume that?

adharkar commented:
Hi @vivek-cumulus @pjdruddy can you review this PR? I have again rebased it.

adharkar commented:
ci:rerun

AnuradhaKaruppiah commented on bgpd/bgp_route.c:
```diff
@@ -9656,6 +9658,27 @@ void route_vty_out_detail(struct vty *vty, struct bgp *bgp, struct bgp_dest *bn,
+			json_object_string_add(json_path, "gatewayIP",
+					       gwip_buf);
+		else
+			vty_out(vty, " Gateway IP %s", gwip_buf);
```
doesn't %pIA work?

adharkar replied:
Will update

adharkar replied:
%pI4 and %pI6 would work. But can we use it for json? In that case we will still need this buf for json.

AnuradhaKaruppiah commented on bgpd/bgp_evpn.c:
```diff
@@ -2480,8 +2508,27 @@ static int install_evpn_route_entry_in_vrf(struct bgp *bgp_vrf,
+
+	/* Gateway IP nexthop should be resolved */
+	if (attr.evpn_overlay.type == OVERLAY_INDEX_GATEWAY_IP) {
+		if (bgp_find_or_add_nexthop(bgp_vrf, bgp_vrf, afi, safi, pi,
```
overlay gw-ip seems like an ideal candidate for the NHG indirection, right? Did you consider that design?

AnuradhaKaruppiah commented on bgpd/bgp_evpn.c:
```diff
@@ -5698,8 +5701,8 @@ int bgp_evpn_local_vni_del(struct bgp *bgp, vni_t vni)
-			   struct in_addr mcast_grp)
-
+			   struct in_addr mcast_grp,
+			   ifindex_t svi_ifindex)
```
Unfortunately this will need to be re-orged again when the vlan-aware-bundle changes come in :(

AnuradhaKaruppiah commented on zebra/zebra_evpn.h:
```diff
@@ -98,6 +98,9 @@ struct zebra_evpn_t_ {
 	struct interface *vxlan_if;
 
+	/* Corresponding SVI interface. */
+	struct interface *svi_if;
```
This will also need to be re-orged later. When vlan-aware-bundle comes in a BD will be represented by a version of zebra_evpn_access_bd.

AnuradhaKaruppiah commented on bgpd/bgp_evpn.h:
```diff
@@ -211,4 +211,9 @@ extern void bgp_evpn_init(struct bgp *bgp);
 extern int bgp_evpn_get_type5_prefixlen(const struct prefix *pfx);
 extern bool bgp_evpn_is_prefix_nht_supported(const struct prefix *pfx);
 extern void update_advertise_vrf_routes(struct bgp *bgp_vrf);
+extern void bgp_evpn_show_remote_ip_hash(struct hash_bucket *bucket,
```
Instead of maintaining a remote_ip_hash in bgp and doing two types of resolution checks (in bgp and zebra) would it make sense to -
1. Setup a NHG for the overlay-index-gw-ip and request zebra to only resolve it via EVPN host routes with an SVI-ifindex (this will also eliminate the need for bgpd to maintain the SVI hash)
2. Add the prefix route via NHG setup in #1. zebra will anyway not install the route till the NHG is resolved.

PS: Maybe I am over-simplifying things? I guess for asym-irb the lack of a host route would make things tricky (putting a dependency on the neigh). If this was already considered during design review and rejected for some reason I would be curious to hear why.

adharkar replied:
As I have explained in the doc, this feature is supported for asymmetric IRB only at this point. Please refer "Topology requirements" in doc/user/bgp.rst

As the draft mentions, overlay index must be resolved using recursive resolution using a type-2 route. So,  if the type-2 route is not present in the BGP VNI table, this prefix should not be marked as valid in BGP as we do not want to advertise the prefix to BGP neighbors in vrf and attract the traffic from the neighbors unnecessarily. I considered doing a neighbor check in Zebra nht, but I decided to add this in BGP as the draft https://tools.ietf.org/html/draft-ietf-bess-evpn-prefix-advertisement-11 specifically mentions "If the RT-5 specifies a GW IP address as the Overlay Index, recursive resolution can only be done if the NVE has received and installed an RT-2 (MAC/IP route) specifying that IP address in the IP address field of its NLRI."

I think nhg won't take care of recursively resolving the gateway IP. Is that correct?

AnuradhaKaruppiah replied:
Yes, I have poured through that draft. I do think NHG is the right answer as it offers the indirection that overlay-index expects. Even if the current NHG support is not sufficient it can be extended.

I understand that this PR is only for aysm just want to be sure that sym doesn't require a complete re-write. I am OK with most of the PR expect for the overlay-index resolution part (which obviously is the crux of the feature!). I think the split between bgpd and zebra needs some analysis so does sym-IRB considerations. I will suggest we have a meeting to discuss that further so we can close on this PR. What do you guys think?

AnuradhaKaruppiah replied:
@adharkar Thanks for the meeting to clarify the design. If we cannot eliminate the need for the remote_macip hash would it make sense to make this processing conditional to the presence of overlay-index config? PS: Not saying it is absolutely necessary; just wanted to see if you had considered it.

adharkar replied:
This would be a good solution. If we know that we are not going to do overlay index processing, we do not need this hash. Let me investigate if this is possible.

AnuradhaKaruppiah commented on zebra/zebra_fpm_netlink.c:
```diff
@@ -199,12 +203,29 @@ static int netlink_route_info_add_nh(struct netlink_route_info *ri,
+			}
 		}
+
+		nhi.encap_info.vxlan_encap.vni = vni;
```
Do you happen to have a dump of the kernel route entries that you can post here?

adharkar replied:
Kernel route entries:

100.0.0.100 via 50.0.1.22 dev br100 proto 186 metric 20 

Note that above change is only for FPM. This encap_info attribute is only used by the FPM here and FPM encodes this even in the pure type-5 case. It is not needed by the kernel data plane. Kernel should do the routing based on the nexthop interface (br100) and nexthop 50.0.1.22.

AnuradhaKaruppiah commented on tests/topotests/bgp-evpn-overlay-index-gateway/PE1/bgp_vni_routes_base.json:
```diff
@@ -0,0 +1,192 @@
+{
```
topotests that cache json output and comp are not reviewable for all practical purposes; as we don't know what exactly the test is trying to validate. I know there are a lot of such tests; but it is not the best precedent IMO.

adharkar replied:
We can discuss this in the meeting. Thanks!

AnuradhaKaruppiah commented on bgpd/bgp_evpn.c:
```diff
@@ -5959,3 +6035,358 @@ bool bgp_evpn_is_prefix_nht_supported(const struct prefix *pfx)
+	 * Gateway IP is resolved by nht over SVI interface.
+	 * Use this SVI to find corresponding EVI(L2 context)
+	 */
+	vpn = bgp_evpn_vni_svi_hash_lookup(bgp_evpn, bnc->nexthop->ifindex);
```
This is assuming that the SVI connected route was used for the nexthop resolution, yes? Will that be the case for both symmetric and asymmetric IRB?

adharkar replied:
As I explained in another comment, this feature supports asymmetric IRB only. But, when we would expand this support to symmetric IRB, nht check can also be expanded for symmetric IRB.

For symmetric IRB, type-2 route corresponding to the overlay index will belong to L3VNI bridge/SVI MAC/IP. At the receiver side, it will be imported into the vrf as a host route and it will be present in zebra as a host route. Thus, When type-5 overlay index is registered with BGP nht, zebra will resolve it over the above host route and will send the reachability info to BGP with L3VNI SVI as the nexthop interface. BGP then can use this interface as a distinguishing factor between overlay index for symmetric IRB vs overlay index for asymmetric IRB. If the overlay index is resolved in zebra over a host route with nexthop interface as L3VNI SVI,  it would mean that a type-2 route is present even for the symmetric IRB overlay index.

AnuradhaKaruppiah reviewed (commented):
I didn't see any changes to path selection. What happens if the same prefix route is rxed from two PEs but the overlay-index-gw-ip is installed as a MAC-IP by only one PE?

adharkar commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -3617,10 +3628,11 @@ DEFUN_HIDDEN (no_bgp_evpn_advertise_vni_subnet,
 DEFUN (bgp_evpn_advertise_type5,
        bgp_evpn_advertise_type5_cmd,
-       "advertise " BGP_AFI_CMD_STR "" BGP_SAFI_CMD_STR " [route-map WORD]",
+       "advertise " BGP_AFI_CMD_STR "" BGP_SAFI_CMD_STR " [gateway-ip] [route-map WORD]",
```
Is redistribute command used to generate type-5 route from a prefix? I do not see any redistribute config command under address-family l2vpn evpn. 

dev(config-router)# address-family l2vpn evpn 
dev(config-router-af)# 
  advertise             Advertise prefix routes
  advertise-all-vni     Advertise All local VNIs
  advertise-default-gw  Advertise All default g/w mac-ip routes in EVPN
  advertise-pip         evpn system primary IP
  advertise-svi-ip      Advertise svi mac-ip routes in EVPN
  autort                Auto-derivation of RT
  default-originate     originate a default route
  disable-ead-evi-rx    Activate PE on EAD-ES even if EAD-EVI is not received
  disable-ead-evi-tx    Don't advertise EAD-EVI for local ESs
  dup-addr-detection    Duplicate address detection
  end                   End current mode and change to enable mode
  exit                  Exit current mode and down to previous mode
  exit-address-family   Exit from Address Family configuration mode
  find                  Find CLI command matching a regular expression
  flooding              Specify handling for BUM packets
  list                  Print command list
  neighbor              Specify neighbor router
  network               Specify a network to announce via BGP
  no                    Negate a command or set its defaults
  output                Direct vtysh output to file
  quit                  Exit current mode and down to previous mode
  rd                    Route Distinguisher
  route-target          Route Target
  use-es-l3nhg          use L3 nexthop group for host routes with ES destination
  vni                   VXLAN Network Identifier

adharkar commented on bgpd/bgpd.h:
```diff
@@ -494,16 +494,18 @@ struct bgp {
-#define BGP_L2VPN_EVPN_ADVERTISE_IPV6_UNICAST		(1 << 2)
-#define BGP_L2VPN_EVPN_DEFAULT_ORIGINATE_IPV4		(1 << 3)
-#define BGP_L2VPN_EVPN_DEFAULT_ORIGINATE_IPV6		(1 << 4)
+#define BGP_L2VPN_EVPN_ADV_IPV4_UNICAST		(1 << 1)
```
Yes. They are mutually exclusive. BGP_L2VPN_EVPN_ADV_IPV4_UNICAST will advertise a ype-5 with no overlay index and BGP_L2VPN_EVPN_ADV_IPV4_UNICAST_GW_IP will advertise a type-5 with gateway IP overlay index. I will add comments for explanation.

adharkar commented:
> I didn't see any changes to path selection. What happens if the same prefix route is rxed from two PEs but the overlay-index-gw-ip is installed as a MAC-IP by only one PE?

In this case, considering both the PEs are advertising different overlay index IPs(For L3 ECMP), overlay index for which type-2 is received is the only valid overlay index. In vrf, only 1 bgp path_info will be valid. Note that a BGP path would take part in the path selection only when the path_info is valid. So, no extra code for the path selection is required. When a type-2 is received for a BGP path with overlay index, nexthop is resolved, path becomes valid and BGP path selection takes place with the regular BGP path selection algorithm.

adharkar commented on bgpd/bgp_evpn.c:
```diff
@@ -5716,13 +5747,53 @@ int bgp_evpn_local_vni_add(struct bgp *bgp, vni_t vni,
+			 */
+			hash_iterate(vpn->remote_ip_hash,
+				(void (*)(struct hash_bucket *,
+				void *))bgp_evpn_remote_ip_hash_unlink_nexthop,
```
Resolved

adharkar commented on bgpd/bgp_evpn.c:
```diff
@@ -5716,13 +5747,53 @@ int bgp_evpn_local_vni_add(struct bgp *bgp, vni_t vni,
+			 */
+			hash_iterate(vpn->remote_ip_hash,
+				(void (*)(struct hash_bucket *,
+				void *))bgp_evpn_remote_ip_hash_link_nexthop,
```
Resolved

adharkar commented on bgpd/bgp_evpn.c:
```diff
@@ -5950,3 +6026,368 @@ bool bgp_evpn_is_prefix_nht_supported(const struct prefix *pfx)
+		return;
+
+	hash_iterate(vpn->remote_ip_hash,
+	(void (*)(struct hash_bucket *, void *))bgp_evpn_remote_ip_hash_free,
```
Resolved

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


---

Pylint found errors in source files changed by this PR:

```
Pylint report for my_frr/tests/topotests/bgp-evpn-overlay-index-gateway/test_bgp_evpn_overlay_index_gateway.py:
************* Module bgp-evpn-overlay-index-gateway.test_bgp_evpn_overlay_index_gateway
my_frr/tests/topotests/bgp-evpn-overlay-index-gateway/test_bgp_evpn_overlay_index_gateway.py:70:0: E0611: No name 'kill_mininet_routers_process' in module 'lib.common_config' (no-name-in-module)

-----------------------------------
Your code has been rated at 9.73/10




```


If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


---

Pylint found errors in source files changed by this PR:

```
Pylint report for my_frr/tests/topotests/bgp-evpn-overlay-index-gateway/test_bgp_evpn_overlay_index_gateway.py:
************* Module bgp-evpn-overlay-index-gateway.test_bgp_evpn_overlay_index_gateway
my_frr/tests/topotests/bgp-evpn-overlay-index-gateway/test_bgp_evpn_overlay_index_gateway.py:70:0: E0611: No name 'kill_mininet_routers_process' in module 'lib.common_config' (no-name-in-module)

-----------------------------------
Your code has been rated at 9.73/10




```


If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


---

Pylint found errors in source files changed by this PR:

```
Pylint report for my_frr/tests/topotests/bgp-evpn-overlay-index-gateway/test_bgp_evpn_overlay_index_gateway.py:
************* Module bgp-evpn-overlay-index-gateway.test_bgp_evpn_overlay_index_gateway
my_frr/tests/topotests/bgp-evpn-overlay-index-gateway/test_bgp_evpn_overlay_index_gateway.py:70:0: E0611: No name 'kill_mininet_routers_process' in module 'lib.common_config' (no-name-in-module)

-----------------------------------
Your code has been rated at 9.73/10




```


If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).
